rfcs/0050-genesis-ledger-export.md

```diff
@@ -0,0 +1,130 @@
+sense to generate that file directly from the old node. Then necessary
+updates can be made to it manually to update various protocol
+constants, and then the new configuration file can be handed over to
+node operators.
+
+## Detailed design
+
+The genesis ledger export is achieved using a GraphQL field named
+`fork_config`. This field, if asked for, contains a new runtime
+configuration, automatically updated with:
+
+* the dump of the current **staged ledger**, which will become the
+genesis ledger for the new network
+* updated values of `Fork_config`, i.e. previous state hash, previous
+blockchain length and previous global slot.
+* updated epoch data, in particular current and next epoch ledger and seed.
+
+**IMPORTANT**: as of now the `genesis_ledger_timestamp` is **not**
+being updated and must be manually set to the right value (which is at
+the moment unknown).
+
+Thus generated configuration can be saved to a file, modified if
+needed and fed directly into a new node, running a different protocol
+version, using `--config-file` flag. As of the moment of writing this,
+`compatible` and `berkeley` branches' configuration files are
+compatible with each other (see: [PR #13768](https://github.com/MinaProtocol/mina/pull/13768)).
+Sadly since then that compatibility has been broken by [PR #14014](https://github.com/MinaProtocol/mina/pull/14014).
+We need to either port this change back to `compatible` or create a
+migration script which will adapt a `mainnet` config file to the
+format required by `berkeley`. The former solution would probably
+be better.
+
+The `fork_config` field has been added to GraphQL in [PR #13787](https://github.com/MinaProtocol/mina/pull/13787).
+
+## Drawbacks
+
+This RFC provides a simple enough procedure to generate the genesis
+ledger for the new network. However, it's not without its problems.
+
+### File size
+
+At the moment the mainnet has more than 100 000 accounts created.
+Each account takes at least 4 lines in the configuration, which adds
+up to around 600kB of JSON data. The daemon can take considerable time
+at startup to parse it and load its contents into memory. If we move
+on with this approach, it might be desirable to make a dedicated
+effort to improving the configuration parsing speed, as these files
+will only grow larger in subsequent hard forks. Alternatively, we
+might want to devise a better (less verbose) storage mechanism for the
+genesis ledger.
+
+### Security concerns
+
+The generated genesis ledger is prone to malevolent manual
+modifications. Beyond containing the hash of the previous ledger, it's
+unprotected from tampering with. However, at the moment there is no
+mechanism which could improve the situation. The system considers
+genesis ledger the initial state of the blockchain, so there is no
+previous state it could refer to. Also, because we dump the **staged
+ledger**, it is never snarked. It can only be verified manually by end
+users, which is cumbersome at best.
```

Comment:
This can be mitigated by providing the program which migrates the captured genesis ledger from the prior chain into the generated genesis ledger for the new chain. With this tool, users can re-execute the program themselves to verify the results, and read the logic of the program to verify it matches what was published by the generating party.

Reply:
I don't think I understand. The genesis ledger file generated by the node is fed directly (except perhaps for manual setting the genesis ledger timestamp) into new nodes. No conversion is required.

Reply:
This is referring to the case where there are ledger changes in the HF in which case you can't directly use the exported ledger. For example, adding/removing a field or any hash function changes changes the ledger hash of the genesis ledger in the HF. So you'd convert the exported ledger to the format that the HF daemon takes. This conversion can be verified by a program that takes the exported ledger, does the conversion itself and verify that the generated ledger has the same hash as the one used in. Assuming the program is simple enough to read and understand how the conversion is taking place. Of course, users can check their accounts to confirm that there are no balance or nonce changes.

Reply:
I see. Adeed in https://github.com/MinaProtocol/mina/pull/14213/commits/6b1e8867e43f9e5b9999979682a11ecd998c5af4.